### PR TITLE
Specify build requirements in pyproject.toml (PEP 517)

### LINKED ISSUE #25193: Pip install of version 0.24 is broken for platforms without wheels
#### Code Sample, a copy-pastable example if possible
Easiest way to reproduce this is in a fresh docker container, as below
```bash
docker run --rm -it alpine:latest
$ apk update
$ apk add build-base python3 python3-dev py3-pip
$ pip3 install pandas
Collecting pandas
  Downloading https://files.pythonhosted.org/packages/81/fd/b1f17f7dc914047cd1df9d6813b944ee446973baafe8106e4458bfb68884/pandas-0.24.1.tar.gz (11.8MB)
    100% |████████████████████████████████| 11.8MB 4.2MB/s 
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 346, in get_provider
        module = sys.modules[moduleOrReq]
    KeyError: 'numpy'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-y3ff_xpr/pandas/setup.py", line 732, in <module>
        ext_modules=maybe_cythonize(extensions, compiler_directives=directives),
      File "/tmp/pip-install-y3ff_xpr/pandas/setup.py", line 475, in maybe_cythonize
        numpy_incl = pkg_resources.resource_filename('numpy', 'core/include')
      File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1136, in resource_filename
        return get_provider(package_or_requirement).get_resource_filename(
      File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 348, in get_provider
        __import__(moduleOrReq)
    ModuleNotFoundError: No module named 'numpy'
```

#### Problem description

Note that pip installing numpy and then pandas does work
```bash
pip3 install numpy
pip3 install pandas
```
But the standalone command is broken.
```bash
pip3 install pandas
```

Unfortunately, this means that a requirements.txt file is insufficient to for setting up a new environment with pandas installed (like in a docker container).

### PR DESCRIPTION
**Pull request by xhochy**
- [x] closes #25193
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Verified this using the Dockefile:

```
FROM alpine:3.8
RUN apk update
RUN apk add g++ libstdc++ python3-dev bash git
RUN pip3 install --upgrade pip

COPY dist/pandas-0.25.0.dev0+86.gba77fb900.tar.gz /
RUN ls -l /
RUN pip3 install -vvv /pandas-0.25.0.dev0+86.gba77fb900.tar.gz
```

### CODE REVIEW

**Comment by xhochy** [comment]
@jorisvandenbossche I hope this works around the issues that were reported before. If you know of others that came with `pyproject.toml` please link me to them. I'll have a look then.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# When changing the version numbers here, also adjust them in `setup.py`
+[build-system]
+requires = ["setuptools", "wheel", "cython >= 0.28.2", "numpy >= 1.12.0"]
```
Comment (jorisvandenbossche) [inline pyproject.toml]: I think we need a fixed old numpy version, depending on the python version. 
At least it's what we did last time (https://github.com/pandas-dev/pandas/commit/77bfe21c7229e724d01721bb84861283baf7e9d3, and it is also what scipy did before they removed it again).
This is because the build happens in an isolated build environment (and with above syntax will always use the latest version), so if your actually installed numpy is then older, it gives problems.

File: setup.py [hunk 2]
```diff
@@ -17,6 +17,15 @@
...
 from setuptools import setup, Command, find_packages
 
+# Taken from https://github.com/jupyterhub/traefik-proxy/
+# commit: 0f16dc307b72e613e71067b6498f82728461434a
+#
+# ensure cwd is on sys.path
+# workaround bug in pip 19.0
+here = os.path.dirname(__file__)
+if here not in sys.path:
+    sys.path.insert(0, here)
```
Comment (jorisvandenbossche) [inline setup.py]: Why is this needed? 
For versioneer?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks! Added some comments.

File: setup.py (hunk 2 above)
Comment (xhochy) [reply setup.py]: Yes, added a comment

File: pyproject.toml (hunk 1 above)
Comment (xhochy) [reply pyproject.toml]: Ah, I'll add that to the `toml`.

File: setup.py (hunk 2 above)
Comment (jorisvandenbossche) [reply setup.py]: BTW, this workaround might not be needed any more, as it should be fixed in setuptools/pip (https://github.com/pypa/pip/pull/6212)

Comment (xhochy) [reply setup.py]: I would keep this in here for a bit as the pip release for that was not done. (From another bug report it seems like it will be in the release coming out this weekend but I won't count on that).

File: pyproject.toml (hunk 1 above)
Comment (xhochy) [reply pyproject.toml]: Actually this fails in my tests as `numpy==1.12.1` doesn't compile on Alpine.

File: setup.py (hunk 2 above)
Comment (jorisvandenbossche) [reply setup.py]: New Pip/Setuptools to fix this have been released in the meantime.

File: pyproject.toml (hunk 1 above)
Comment (jorisvandenbossche) [reply pyproject.toml]: > Actually this fails in my tests as `numpy==1.12.1` doesn't compile on Alpine.

Is that a known issue with numpy? 
But not really sure how to deal with that. I think we need the pins here for the general case. It might be, that if our setup does not work for a specific use case / platform, you will need to use a `--no-build-isolation`

File: setup.py (hunk 2 above)
Comment (xhochy) [reply setup.py]: Removed this again

File: pyproject.toml (hunk 1 above)
Comment (xhochy) [reply pyproject.toml]: Yes, this is a known issue with NumPy and was fixed in the 1.13 series. I have now pinned the Python 3.6 build also to 1.13.1 and added an explanatory comment. As Alpine+Python3.6 was the initial origin for tackling the `pyproject.toml` again, I made this change.

**Comment by jorisvandenbossche** [comment]
> Yes, this is a known issue with NumPy and was fixed in the 1.13 series. I have now pinned the Python 3.6 build also to 1.13.1 and added an explanatory comment. As Alpine+Python3.6 was the initial origin for tackling the `pyproject.toml` again, I made this change.

This might ensure it works on alpine+python3.6, but then can give errors on other platforms in case you would eg do `pip install pandas numpy=1.12 --no-binary :all:` (or without the no-binary on platforms that don't have wheels, but where numpy 1.12 works).

I am not sure what the best way to go is here (although I am inclined to keep it at the lowest we support (1.12), so it works when numpy 1.12 works on the specific platform). 
But, it is a problem that many other packages also will encounter when adding (back) pyproject.toml file, like scipy, scikit-learn, ... So it might be interesting to know what they would do / to be consistent (as often those packages will be installed together)

cc @rgommers (since you were involved in the pyproject.toml file in scipy)

**Comment by rgommers** [comment]
Pip support should finally be good enough now with pip 19, we're adding it back in SciPy soon: https://github.com/scipy/scipy/issues/8734

> (although I am inclined to keep it at the lowest we support (1.12), so it works when numpy 1.12 works on the specific platform).

Yes, this is the only correct way to do it, you have to build against the lowest supported numpy version for each Python version. If that lowest version varies (e.g. Python 3.7 + numpy 1.12.x won't work), then specify multiple versions like so: https://github.com/PyWavelets/pywt/commit/5e6d53a75b5c682ef1b8e40cefc57df7e7e2b399#diff-522adf759addbd3b193c74ca85243f7d

**Comment by jorisvandenbossche** [comment]
> Yes, this is the only correct way to do it, you have to build against the lowest supported numpy version for each Python version.

Yes, that's what we are doing (and did before as well before we removed the pyproject.toml again .. just like scipy did as well :)). 
But the question was more specifically about the situation for alpine linux, where apparently, numpy 1.12 doesn't work for python 3.6. So pinning to numpy 1.12, will mean that people on alpine (where typically no wheels are available so this pyproject.toml is of relevance) will still need to do a `--no-build-isolation` or `--no-use-pep517` to be able to install pandas. 
But I suppose there is not really a way around that?

**Comment by jreback** [comment]
maybe soln is to bump numpy requirement to 1.13? or higher

**Comment by xhochy** [comment]
Is is really that common that people are using Pandas with a numpy version older than a year? Updating numpy has been quite smooth in the last years, so I would have expected that this is one of the simpler migrations.

**Comment by jreback** [comment]
agree if bumping helps here let’s do it
only downside is that then this patch must be for 0.25 but that should be soonish anyhow

**Comment by jorisvandenbossche** [comment]
> only downside is that then this patch must be for 0.25 but that should be soonish anyhow

I would keep this patch for 0.25 anyway. We don't want to experiment with pyproject.toml in a bug fix release (it should not give that much troubles as last time we added it, but still)

**Comment by rgommers** [comment]
> But I suppose there is not really a way around that?

Indeed there isn't. There are no features of `pyproject.toml` that allow you to make the numpy version that you build against depend on Alpine/MUSL vs other linux distros.

> Is is really that common that people are using Pandas with a numpy version older than a year?

Yes it is. I'm quite sure that that's more common than people using non-manylinux-compatible distros.

> maybe soln is to bump numpy requirement to 1.13? or higher

In this case, SciPy is anyway bumping to 1.13.3 as minimum numpy version for the next release. So bumping won't impact many users. So +1 for that solution.

**Comment by xhochy** [comment]
The `pip` version on Azure is not yet `19.0.2`. I wonder whether it would be better to have the workaround in place to support the missing current directory in the `sys.path` rather than updating `pip` in the CI jobs.

**Comment by TomAugspurger** [comment]
Does this fully close 
https://github.com/pandas-dev/pandas/issues/25193, or is there an additional issue there with `cythonize` being called unconditionally?

**Comment by jorisvandenbossche** [comment]
If we are fine with cython being a build dependency (also for sdists, as pyproject.toml cannot really distinguish from source vs from sdist builds), then the unconditional call to cythonize should not really be a problem. 

But as I said already before, I don't think this PR closes #25193 *for 0.24.2* (we don't want to start using pyproject.toml in a bug fix release, and also the numpy version increase should go in 0.25).

**Comment by jorisvandenbossche** [comment]
>  I wonder whether it would be better to have the workaround in place to support the missing current directory in the `sys.path` rather than updating `pip` in the CI jobs.

It's not that urgent to merge this I think, so we might still be able to remove it before that. 
But, that said, we should maybe also think about adding `versioneer` to the build-system `requires`, instead of relying on the one included in the source, if we want to prevent that we need to have the current directory on the python path.

**Comment by h-vetinari** [comment]
Bumping numpy comes with a bit more complexity (esp. re: compat code) - I've started a PR dedicated to just that in #25554.

**Comment by xhochy** [comment]
@h-vetinari Thanks for taking care of the NumPy bump. I hope I can spend some time on this again once the bump is merged.

**Comment by jorisvandenbossche** [comment]
Yes, it is a good idea to do the numpy version bump in a separate PR.

**Comment by h-vetinari** [comment]
> @xhochy: @h-vetinari Thanks for taking care of the NumPy bump. I hope I can spend some time on this again once the bump is merged.

The numpy bump has been merged a few days ago. :)

**Comment by jreback** [comment]
@xhochy can you merge master. we are now in 1.13 min, so you can remove a lot of these changes.

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
https://travis-ci.org/pandas-dev/pandas/jobs/529095120

```
ERROR: Error installing 'file:///home/travis/build/pandas-dev/pandas': editable mode is not supported for pyproject.toml-style projects. pip is processing this project as pyproject.toml-style because it has a pyproject.toml file. Since the project has a setup.py and the pyproject.toml has no "build-backend" key for the "build_system" value, you may pass --no-use-pep517 to opt out of pyproject.toml-style processing. See PEP 517 for details on pyproject.toml-style projects.
The command "ci/setup_env.sh" failed and exited with 1 during .
```

seems like https://github.com/pypa/pip/issues/6434

**Comment by jorisvandenbossche** [comment]
Yes, there was indeed a regression with the latest pip 19.1.0 for projects using editable installs and already having pyproject.toml file (although I think pip 19.1.1 which was just released should mostly fix that)

We could choose to postpone adopting a pyproject.toml somewhat more until the dust settles around it (I think people are working on a PEP update to include editable installs in PEP517/518, but that will take some more time).

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -415,6 +415,7 @@ Reshaping
 - Bug in :func:`pandas.cut` where large bins could incorrectly raise an error due to an integer overflow (:issue:`26045`)
 - Bug in :func:`DataFrame.sort_index` where an error is thrown when a multi-indexed DataFrame is sorted on all levels with the initial level sorted last (:issue:`26053`)
 - Bug in :meth:`Series.nlargest` treats ``True`` as smaller than ``False`` (:issue:`26154`)
+- Bug in :func:`factorize` when passing an ``ExtensionArray`` with a custom ``na_sentinel`` (:issue:`25696`).
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.25.0.rst]: some merge left-overs?

File: setup.py [hunk 4]
```diff
@@ -17,6 +17,16 @@
 from distutils.version import LooseVersion
 from setuptools import setup, Command, find_packages
 
+# Taken from https://github.com/jupyterhub/traefik-proxy/
+# commit: 0f16dc307b72e613e71067b6498f82728461434a
+#
+# ensure cwd is on sys.path
+# workaround bug in pip 19.0
```
Comment (jorisvandenbossche) [inline setup.py]: I think all of this can be removed now, see also discussion below

**Comment by TomAugspurger** [comment]
Is this a blocker for 0.25.0, or were all the blocking issue resolved elsewhere?

**Comment by jorisvandenbossche** [comment]
I don't think this is urgent (eg https://github.com/pandas-dev/pandas/issues/25193 which triggered, you already solved that on master)

**Comment by jreback** [comment]
closing; we need to see if we actually need this.
